tests/document/test_instance.py

```diff
@@ -1457,6 +1457,53 @@ def test_inserts_if_you_set_the_pk(self):
 
         assert 2 == self.Person.objects.count()
 
+    def test_save_upsert_false_doesnt_insert_when_deleted(self):
+        class Person(Document):
+            name = StringField()
+
+        Person.drop_collection()
+
+        p1 = Person(name="Wilson Snr")
+        p1.save()
+        p2 = Person.objects().first()
+        p1.delete()
+        p2.name = " Bob Snr"
+        p2.save(upsert=False)
+
+        assert Person.objects.count() == 0
+
+    def test_save_upsert_true_inserts_when_deleted(self):
+        class Person(Document):
+            name = StringField()
+
+        Person.drop_collection()
+
+        p1 = Person(name="Wilson Snr")
+        p1.save()
+        p2 = Person.objects().first()
+        p1.delete()
+        p2.name = "Bob Snr"
+        p2.save(upsert=True)
+
+        assert Person.objects.count() == 1
+
+    def test_save_upsert_null_inserts_when_deleted(self):
+        # probably want to remove this as this is bad but preserved for backwards compatibility
+        # see https://github.com/MongoEngine/mongoengine/issues/564
+        class Person(Document):
+            name = StringField()
+
+        Person.drop_collection()
+
+        p1 = Person(name="Wilson Snr")
+        p1.save()
+        p2 = Person.objects().first()
+        p1.delete()
+        p2.name = "Bob Snr"
+        p2.save(upsert=None)  # default if you dont pass it
+
+        assert Person.objects.count() == 1
+
```

Comment:
I believe we should have 1 (perhaps more) test that ensures how providing both save_condition & upsert would behave.
Depending on how we resolved some of the comments I made in this PR, we'll probably need a few more tests as well